core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -30,96 +54,209 @@
 
     public static final String NAME = "terms";
 
-    static final TermsQueryBuilder PROTOTYPE = new TermsQueryBuilder(null, (Object) null);
+    static final TermsQueryBuilder PROTOTYPE = new TermsQueryBuilder(null);
 
-    private final String name;
-
-    private final Object values;
+    public static final boolean DEFAULT_DISABLE_COORD = false;
 
+    private final String fieldName;
+    private final List<Object> values = new ArrayList<>();
     private String minimumShouldMatch;
+    private boolean disableCoord = DEFAULT_DISABLE_COORD;
+    private TermsLookupBuilder termsLookup;
+
+    public static final class TermsLookupBuilder implements Writeable<TermsLookupBuilder>, ToXContent {
+        static final TermsLookupBuilder PROTOTYPE = new TermsLookupBuilder();
+
+        private String index;
+        private String type;
+        private String id;
+        private String path;
+        private String routing;
+
+        public TermsLookupBuilder() {
+        }
+
+        public TermsLookupBuilder(@Nullable String index, String type, String id, String path,
+                                  @Nullable String routing) {
+            this.index = index;
+            this.type = type;
+            this.id = id;
+            this.path = path;
+            this.routing = routing;
+        }
+
+        public TermsLookup toTermsLookup(QueryShardContext context) {
```

Comment:
I thought we had agreed to have this inner class to replace the equivalent member variables. There is very little difference between TermsLookup and TermsLookupBuilder though now that I think of it. TermsLookup does take a QueryParseContext though. I say we refactor TermsLookup into TermsLookupBuilder when we refactor TermsLookupQueryBuilder.